File: daiphat-fe/src/client/services/transactionService.ts
import { apiApp } from '../../api';
import {
    CancelPaymentRequest,
    PaymentResult,
    PendingPaymentCountdownResult,
    ProcessPaymentRequest
} from '../../types/transaction.type';
import { ApiResponse } from '../../admin/config/type';

const BASE_URL = '/transactions';

export const transactionService = {
    /**
     * Generates the payment checkout URL
     */
    processPayment: async (orderId: string, data: ProcessPaymentRequest): Promise<ApiResponse<PaymentResult>> => {
        const response = await apiApp.post(`${BASE_URL}/${orderId}/payment`, data);
        return response.data;
    },

    cancelPayment: async (orderId: string, data: CancelPaymentRequest): Promise<ApiResponse<any>> => {
        const response = await apiApp.post(`${BASE_URL}/${orderId}/payment/cancel`, data);
        return response.data;
    },

    getPendingPaymentCountdown: async (orderId: string): Promise<ApiResponse<PendingPaymentCountdownResult>> => {
        const response = await apiApp.get(`${BASE_URL}/${orderId}/payment/countdown`);
        return response.data;
    },
    
    /**
     * Gets available transaction types
     */
    getTransactionTypes: async (): Promise<ApiResponse<any[]>> => {
        const response = await apiApp.get(`${BASE_URL}/types`);
        return response.data;
    }
};

File: daiphat_mobile/lib/src/features/checkout/data/transaction_service.dart
import 'package:daiphat_mobile/src/shared/network/api_client.dart';
import '../models/transaction_type.dart';

class TransactionService {
  static const _baseTransactions = '/transactions';

  final ApiClient _apiClient;

  TransactionService(this._apiClient);

  Future<List<EnumOption>> getTransactionTypes() async {
    final response = await _apiClient.get('$_baseTransactions/types');
    final data = response['data'];
    if (data is List) {
      return data
          .map((e) => EnumOption.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    return [];
  }

  Future<PaymentResult> processPayment({
    required String orderId,
    required ProcessPaymentRequest request,
  }) async {
    final response = await _apiClient.post(
      '$_baseTransactions/$orderId/payment',
      data: request.toJson(),
    );
    final data = response['data'] as Map<String, dynamic>;
    return PaymentResult.fromJson(data);
  }

  Future<PendingPaymentCountdownResult> getPendingPaymentCountdown(
    String orderId,
  ) async {
    final response = await _apiClient.get(
      '$_baseTransactions/$orderId/payment/countdown',
    );
    final data = response['data'] as Map<String, dynamic>;
    return PendingPaymentCountdownResult.fromJson(data);
  }
}
